packages/modules/pricing/src/services/pricing-module.ts

```diff
@@ -238,67 +255,115 @@ export default class PricingModuleService
-          let originalPrice: PricingTypes.CalculatedPriceSetDTO = defaultPrice
+    const calculatedPrices: PricingTypes.CalculatedPriceSet[] = (
+      await promiseAll(
+        pricingFilters.id.map(
+          async (
+            priceSetId: string
+          ): Promise<PricingTypes.CalculatedPriceSet | null> => {
+            // This is where we select prices, for now we just do a first match based on the database results
+            // which is prioritized by rules_count first for exact match and then default_priority of the rule_type
+
+            const prices = pricesSetPricesMap.get(priceSetId) || []
+            const priceListPrice = prices.find((p) => p.price_list_id)
+            const defaultPrice = prices?.find((p) => !p.price_list_id)
+            if (!prices.length || !defaultPrice) {
+              return null
+            }
 
-          if (priceListPrice) {
-            calculatedPrice = priceListPrice
+            let calculatedPrice: PricingTypes.CalculatedPriceSetDTO =
+              defaultPrice
+            let originalPrice: PricingTypes.CalculatedPriceSetDTO = defaultPrice
+            if (priceListPrice) {
+              calculatedPrice = priceListPrice
 
-            if (priceListPrice.price_list_type === PriceListType.OVERRIDE) {
-              originalPrice = priceListPrice
+              if (priceListPrice.price_list_type === PriceListType.OVERRIDE) {
+                originalPrice = priceListPrice
+              }
             }
-          }
 
-          return {
-            id: priceSetId,
-            is_calculated_price_price_list: !!calculatedPrice?.price_list_id,
-            calculated_amount: parseInt(calculatedPrice?.amount || "") || null,
-
-            is_original_price_price_list: !!originalPrice?.price_list_id,
-            original_amount: parseInt(originalPrice?.amount || "") || null,
-
-            currency_code: calculatedPrice?.currency_code || null,
-
-            calculated_price: {
-              id: calculatedPrice?.id || null,
-              price_list_id: calculatedPrice?.price_list_id || null,
-              price_list_type: calculatedPrice?.price_list_type || null,
-              min_quantity:
-                parseInt(calculatedPrice?.min_quantity || "") || null,
-              max_quantity:
-                parseInt(calculatedPrice?.max_quantity || "") || null,
-            },
-
-            original_price: {
-              id: originalPrice?.id || null,
-              price_list_id: originalPrice?.price_list_id || null,
-              price_list_type: originalPrice?.price_list_type || null,
-              min_quantity: parseInt(originalPrice?.min_quantity || "") || null,
-              max_quantity: parseInt(originalPrice?.max_quantity || "") || null,
-            },
+            // We use the price rules to get the right preferences for the price
+            const priceRulesForPrices = await this.priceRuleService_.list(
```

Comment:
Can we bring this .list outside of the loop to avoid multiple db calls?

Reply:
Yes let's do it, it will require some more cleanup so I thought I leave it for now, but I think we should